src/components/select/filterable-select/filterable-select.component.js

```diff
@@ -0,0 +1,293 @@
+    const isDeleteEvent = event.nativeEvent.inputType === 'deleteContentBackward'
+      || event.nativeEvent.inputType === 'deleteContentForward'
+      || event.nativeEvent.inputType === 'delete';
+    updateValues(newValue, isDeleteEvent);
+    setOpen(true);
+  }, [setOpen, updateValues]);
+
+  const handleTextboxKeydown = useCallback(((event) => {
+    const isHomeOrEndKey = Events.isHomeKey(event) || Events.isEndKey(event);
+    const isUpOrDownKey = Events.isDownKey(event) || Events.isUpKey(event);
+
+    if (isUpOrDownKey || isHomeOrEndKey) {
+      event.preventDefault();
+      setOpen(true);
+    }
+  }), [setOpen]);
+
+  const handleGlobalClick = useCallback((event) => {
+    const notInContainer = containerRef.current && !containerRef.current.contains(event.target);
+    const notInList = listboxRef.current && !listboxRef.current.contains(event.target);
+
+    if (notInContainer && notInList) {
+      setMatchingText(selectedValue);
+      setOpen(false);
+    }
+  }, [setMatchingText, setOpen, selectedValue]);
+
+  useEffect(() => {
+    const newValue = value || defaultValue;
+    const message = 'Input elements should not switch from uncontrolled to controlled (or vice versa). '
+    + 'Decide between using a controlled or uncontrolled input element for the lifetime of the component';
+    invariant(isControlled.current === (value !== undefined), message);
+
+    setSelectedValue(newValue);
+    setHighlightedValue(newValue);
+    setMatchingText(newValue);
+  }, [value, defaultValue, setMatchingText]);
+
+  useEffect(() => {
+    const clickEvent = 'click';
+
+    document.addEventListener(clickEvent, handleGlobalClick);
+
+    return function cleanup() {
+      document.removeEventListener(clickEvent, handleGlobalClick);
+    };
+  }, [handleGlobalClick]);
+
+  useEffect(() => {
+    const textStartsWithFilter = textValue.toLowerCase().startsWith(filterText.toLowerCase());
+
+    if (textboxRef && textValue.length > filterText.length && textStartsWithFilter) {
+      textboxRef.selectionStart = filterText.length;
+    }
+  }, [textValue, filterText, textboxRef]);
+
+  function handleTextboxClick(event) {
+    const elementType = event.target.attributes['data-element'];
+    const isInputClicked = elementType && elementType.value === 'input';
+
+    if (onClick) {
+      onClick(event);
+    }
+
+    if (!elementType || !isInputClicked) {
+      setOpenState((isAlreadyOpen) => {
+        if (isAlreadyOpen) {
+          return false;
+        }
+
+        if (onOpen) {
+          onOpen();
+        }
+
+        return true;
+      });
+    }
+  }
+
+  function handleTextboxFocus(event) {
+    if (onFocus) {
+      onFocus(event);
+    }
+  }
+
+  function onSelectOption(optionData) {
+    const { text, value: newValue, selectionType } = optionData;
+
+    if (!isControlled.current) {
+      setSelectedValue(newValue);
+      setTextValue(text);
+    }
+
+    if (onChange) {
+      onChange(createCustomEvent(newValue));
```

Comment:
I know we raised an RFC about this, primarily for difficult components that store multiple values.

In this example, is the value the same that is in the DOM? If so we can just pass the ref `onChange(textboxRef.current)`